### PR TITLE
Base GLWidget on GLArea instead of DrawingArea

### PR DESCRIPTION
**Pull request by Nihlus**
### Purpose of this PR

This PR reworks the GLWidget such that it is no longer based off of DrawingArea - instead, it is based on GLArea, which is a dedicated OpenGL widget for GTK.

This allows OpenTK to delegate all platform-specific context initialization to GTK, thereby making a swath of code in GLWidget obsolete and removable. Furthermore, it solves the issue where depth buffers would be broken on NVIDIA drivers on Linux (see #464) due to a window visual mismatch with the context visual.

There are some drawbacks, however. Firstly, the public API of the widget has significantly changed - mostly for the better, but a little for the worse. Secondly, the widget is somewhat less configurable than before, due the to public API available in GLArea and its C# bindings being more limited than what we were doing by directly drawing to the window. This only affects the X11 platform, and not the macOS or Windows platforms.

Using GLArea also allows far better integration into GTK as a whole, and enables more advanced compositing features such as rendering other widgets on top of an OpenGL viewport. Additionally, all of the platform-specific initialization code was rendered obsolete and redundant by this change, and has been removed.

The full documentation for GLArea is available [here](https://developer.gnome.org/gtkmm/stable/classGtk_1_1GLArea.html), and the GdkGLContext documentation is available [here](https://developer.gnome.org/gtkmm/stable/classGdk_1_1GLContext.html).

## Removals
* GTK2 support has been removed. Use a previous version of GLWidget if this is required.
* `GlVersionMajor` has been removed. Use the constructor parameter instead, or `Set/GetRequiredVersion`.
* `GlVersionMinor` has been removed. Use the constructor parameter instead, or `Set/GetRequiredVersion`.
* `SingleBuffer` has been removed. `GLArea` is double-buffered by default; Use `Set/GetDoubleBuffered` instead.
* `ColorBPP` has been removed. `GLArea` selects the best colour available instead. Use `HasAlpha` to enable or disable the alpha channel.
* `AccumulatorBPP` has been removed. `GLArea` does not support an accumulator buffer.
* `DepthBPP` has been removed. `GLArea` selects the best depth available instead. Use `HasDepthBuffer` to enable or disable the buffer.
* `StencilBPP` has been removed. `GLArea` selects the best stencil available instead. Use `HasStencilBuffer` to enable or disable the buffer.
* `Samples` has been removed. `GLArea` does not support a configurable sample count.
* `Stereo` has been removed. `GLArea` does not support stereoscopic rendering.
* The `RenderFrame` event has been removed. Use the `Render` event provided by `GLArea` instead.
* `GLWidget::GraphicsContextFlags` has been removed. It is replaced by a get-only property (`GLWidget::ContextFlags`).

## Behaviour changes
* Rendering to the widget no longer requires the user to manually call `GraphicsContext::MakeCurrent` or `GraphicsContext::SwapBuffers` on the context. Instead, the `Render` event is raised by the widget, and any code inside it is guaranteed to have a context current on the thread, and any results from the drawing code inside will be swapped to the widget's framebuffer at the end of the signal.
* Rendering should no longer be done through attaching an idle timer and running the event manually. Instead, render events should be manually queued at the end of each frame by using `GLArea::QueueRender`.
* Frame time delta is now measured by GDK. GDK synchronizes to the monitor's refresh rate by default, and manually measuring your frame time will produce incorrect results. Instead, a new property (`GLWidget::DeltaTime`) is provided which is guaranteed to contain the actual frame delta in the `GLWidget::Render`, `GLWidget::GraphicsContextInitialized`, and `GLWidget::Initialized`.
* `GLArea` does not explicitly support contexts below 3.2, due to the deprecation of the FFP API. If one below 3.2 is requested, one of at least version 3.2 will be returned and a warning will be issued to `stdout`.
* Buffer BPP values are no longer explicitly set in the widget. Instead, a passed-in `OpenTK::GraphicsMode` will be used to hint the `GLArea` base about which buffers should be created, and the best available values will be picked.
* The widget respects `GraphicsContextFlags::Debug` and `GraphicsContextFlags::ForwardCompatible` as before. However, error checking is also enabled for contexts in widgets with the Debug flag set.

### Other changes
* The widget no longer performs any platform-reliant initialization by itself, instead relying on the backend GDK/GTK implementation.

### Testing status

* Extensively tested with a nontrivial application on Linux Mint 18.2, using the 384.69 NVIDIA drivers. Applications mentioned in #464 function normally after this patch.
* No testing on Windows systems. However, all platform-specific initialization is now delegated to GDK, and it is a reasonable assumption that it should not be affected.
* No testing on Mac systems. However, all platform-specific initialization is now delegated to GDK, and it is a reasonable assumption that it should not be affected.

### CODE REVIEW

File: src/OpenTK/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -9,6 +10,9 @@
 [assembly: AssemblyFileVersionAttribute("3.0.0")]
 [assembly: CLSCompliantAttribute(true)]
 [assembly: AssemblyCopyrightAttribute("Copyright (c) 2006 - 2016 Stefanos Apostolopoulos <stapostol@gmail.com> for the Open Toolkit library.")]
+
+[assembly: InternalsVisibleTo("OpenTK.GLWidget, PublicKey=0024000004800000940000000602000000240000525341310004000011000000a3e05aafb87f71fb8fd02b512707f289c12341de98c6ce2ec1494b71c20cb2032cbd65d091b447df3ec772257efb9fa3591201937890e067da1d29a339948a12b29c2847a787cc9fbef2a3bf78267026e85f36aab827228df4bb580e3ae0599ade036f0a97a0e6982f5406d98d114455f332350f6d8369db655e9c3e7976c2c8")]
```
Comment (daerogami) [inline src/OpenTK/Properties/AssemblyInfo.cs]: I recall there being a discussion of modularity where the project will have a "core" and the additional features could be included as needed.

If that's still a goal, this strong coupling may be a cause for concern.

Comment (Nihlus) [reply src/OpenTK/Properties/AssemblyInfo.cs]: Right. In this particular case, the widget project is still completely modular, but will not function without the main OpenTK library. The main OpenTK library does not depend on the GLWidget project by allowing it to access internals - it's just a bit of metadata. 

That being said, with some additional changes I've not yet pushed, marking it with InternalsVisibleTo is no longer required, and will be removed.

**Comment by varon** [comment]
Are we going to hold off until 4.0.0 with this? I don't really mind merging for 3.0.0 either.

**Comment by Nihlus** [comment]
@varon I think this is worth merging in for 3.0.0. We can drop GTK2 in the same go, although we'll have to be clear that GLWidget will only support GTK >= 3.16 after the merge. Users still using GTK2 are, of course, still able to stick on the old package version.

**Comment by Nihlus** [comment]
@varon @daerogami I've now pushed the final changes. This brings some more API changes, removes GTK2 support, and reverts the InternalsVisibleTo change. Furthermore, the Debug and ForwardCompatible flags are supported again.

**Pull request merged**